tests/models/test_jsonapi.py

```diff
@@ -1,6 +1,57 @@
-from optimade.models.jsonapi import Error
+from pydantic import ValidationError
+import pytest
+from optimade.models.jsonapi import Error, ToplevelLinks
 
 
 def test_hashability():
     error = Error(id="test")
     assert set([error])
+
+
+def test_toplevel_links():
+    """Test top-level links responses as both URLs and Links objects,
+    and check that any arbitrary key is allowed as long as the value
+    can be validated as a URL or a Links object too.
+
+    """
+    test_links = {
+        "first": {"href": "http://example.org/structures?page_limit=3&page_offset=0"},
```

Comment:
```suggestion
        "first": "http://example.org/structures?page_limit=3&page_offset=0",
```

Reply:
The combination of AnyUrl/Links is tested for each field inside the loop. Likewise below, it feels a bit silly testing that an pydantic `Optional` field is allowed to be None...

Reply:
Fair enough. I thought you'd just set up the mock links response object and then pass it as a whole to `TopLevelLinks` for validation. It would be good to have it represent all the different ways of representing links then, all in one.